File: examples/javascript/TypeORM/repositories.ts
import { EntityRepository, Repository } from 'typeorm';
import { Post } from '../entities/post.entity';
import { PostSection } from '../entities/post-section.entity';

@EntityRepository(Post)
export class PostRepository extends Repository<Post> {

  async findValidPosts(
    blogId: number,
    authorIds?: number[],
    sectionCode?: string,
    offset = 0,
    limit = 10,
    allowedStatuses?: string[],  // ex: ['검토중', '검토완료', '게시됨']
  ): Promise<Post[]> {
    const query = this.createQueryBuilder('post')
        .leftJoinAndSelect('post.postSections', 'section');

    // 조건 붙이기
    if (blogId) {
        query.andWhere('post.blogId = :blogId', { blogId: blogId });
    }

    if (authorIds && authorIds.length > 0) {
        query.andWhere('post.authorId IN (:...authorIds)', { authorIds: authorIds });
    }

    if (sectionCode) {
      const subQuery = query.subQuery()
        .select('section.postId')
        .from(PostSection, 'section')
        .where('section.sectionCode = :sectionCode', { sectionCode })
        .getQuery();

      if (offset > 0) {
        query.andWhere(`post.id NOT IN ${subQuery}`);
      } else {
        query.andWhere(`post.id IN ${subQuery}`);
      }
    }

    if (allowedStatuses && allowedStatuses.length > 0) {
      query.andWhere('section.permitComment = true')
           .andWhere('section.sectionStatus IN (:...allowedStatuses)', { allowedStatuses });
    }

    query.orderBy('post.postedAt', 'DESC')
         .skip(offset)
         .take(limit);

    return query.getMany();
  }
}

File: examples/python/SQLAlchemy/repositories.py
class PostRepository:
    def __init__(self, db: Session):
        self.db = db

    def get_base_query(self):
        return self.db.query(Post)

    def filter_by_blog(self, query, blog_id: int):
        return query.filter(Post.blog_id == blog_id)

    def filter_by_authors(self, query, author_ids: list[int]):
        return query.filter(Post.author_id.in_(author_ids)) if author_ids else query

    def filter_by_section_code(self, query, section_code: str, offset: int):
        if not section_code:
            return query
        subq = self.db.query(PostSection.post_id).filter(PostSection.section_code == section_code)
        return query.filter(~Post.id.in_(subq) if offset else Post.id.in_(subq))

    def filter_by_section_status(self, query, allowed_statuses: list[int]):
        return query.filter(
            exists().where(
                (PostSection.post_id == Post.id)
                & (PostSection.permit_comment == True)
                & (PostSection.section_status.in_(allowed_statuses))
            )
        )
